### ISSUE TITLE
image do not show up , some times 

### ISSUE THREAD

**Issue by utkarshGithub** [noties/Markwon#102]
Sometimes images do not render at all, there is a OBJ image appearing there, what can be done and  how to use a place holder for images

**Comment by noties** [comment]
Hello @utkarshGithub !

Let me start with the placeholder. It's possible, for example like this (please note that this is for upcoming `3.0.0` release, but `2.x.x` should be almost the same, the only difference is that you have to create own instance `SpannableFactory` and override `image` method):

```java
final Markwon markwon = Markwon.builder(context)
        .usePlugin(ImagesPlugin.createWithAssets(context))
        .usePlugin(new AbstractMarkwonPlugin() {
            @Override
            public void configureSpansFactory(@NonNull MarkwonSpansFactory.Builder builder) {
                builder.setFactory(Image.class, new SpanFactory() {
                    @Override
                    public Object getSpans(
                            @NonNull MarkwonConfiguration configuration,
                            @NonNull RenderProps props) {

                        // create an instance of AsyncDrawable
                        final AsyncDrawable drawable = new AsyncDrawable(
                                ImageProps.DESTINATION.require(props),
                                configuration.asyncDrawableLoader(),
                                configuration.imageSizeResolver(),
                                ImageProps.IMAGE_SIZE.get(props));

                        // setResult immediately, this result will be treated as a placeholder
                        // and will be displayed whilst image is loading
                        drawable.setResult(new PlaceholderDrawable(
                                context.getResources().getDrawable(R.drawable.ic_play_circle_filled_18dp_white),
                                Color.RED));

                        return new AsyncDrawableSpan(
                                configuration.theme(),
                                drawable,
                                AsyncDrawableSpan.ALIGN_BOTTOM,
                                ImageProps.REPLACEMENT_TEXT_IS_LINK.get(props, false)
                        );
                    }
                });
            }
        })
        .build();
```

Although you can use whatever Drawable as a placeholder, including one bundled with your application in resources, I would advice to create an own instance of it. The main reason is: different images will have different sizes, so your _raw_ drawable asset will be scaled to fit the available Canvas. But if you would use something like this:

```java
public class PlaceholderDrawable extends Drawable {

    private final Drawable icon;
    private final Paint paint;

    public PlaceholderDrawable(@NonNull Drawable icon, @ColorInt int backgroundColor) {
        this.icon = icon;

        final Rect rect = new Rect(0, 0, icon.getIntrinsicWidth(), icon.getIntrinsicHeight());

        // do not forget to init with intrinsic width & height
        this.icon.setBounds(rect);

        // also, do not forget to set own bounds to be at least like the icon ones
        setBounds(rect);

        this.paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        this.paint.setStyle(Paint.Style.FILL);
        this.paint.setColor(backgroundColor);
    }

    @Override
    public void draw(@NonNull Canvas canvas) {
        final int save = canvas.save();
        try {

            final Rect iconBounds = icon.getBounds();
            final Rect bounds = getBounds();

            // calculate translation values to center our icon
            // NB this calculations expect icon to be always smaller than actual bounds,
            // which _might_ not always be the case
            final float left = ((float) bounds.width() - iconBounds.width()) / 2.F;
            final float top = ((float) bounds.height() - iconBounds.height()) / 2.F;

            canvas.drawRect(bounds, paint);

            canvas.translate(left, top);
            icon.draw(canvas);

        } finally {
            canvas.restoreToCount(save);
        }
    }

    @Override
    public void setAlpha(int alpha) {
        // no op
    }

    @Override
    public void setColorFilter(@Nullable ColorFilter colorFilter) {
        // no op
    }

    @Override
    public int getOpacity() {
        return PixelFormat.OPAQUE;
    }
}
```

Then your placeholder will fit the available Canvas and draw a proper background plus centering actual placeholder icon (keeping the image ratio).

Now, you see the ￼(obj) when image is loading or failed loading because for an image to have a proper span, it must have at least one character. I've used ￼(obj) because it is what Android uses in `Html.fromHtml`, although I do agree that it's not the best option. Actually, I'm going to add a placeholder option to image loading functionality for the next `3.0.0` release. Meanwhile, you still can use `errorDrawable` when configuring `AsyncDrawable.Loader` to display when loading had failed.

So. Why some images fail? It's hard to tell. And there could be a million of reasons. The bad part is, current implementation gives no clue what had happened (no error logs, nothing). I'm going to work on that and hope that this change will land in the next release.

🙌

**Comment by noties** [comment]
And actually, applying placeholder _currently_ can be achieved in a bit different way (omitting some image handling details):

```java
.usePlugin(new AbstractMarkwonPlugin() {
    @Override
    public void configureSpansFactory(@NonNull MarkwonSpansFactory.Builder builder) {
        final SpanFactory original = builder.getFactory(Image.class);
        if (original != null) {
            builder.setFactory(Image.class, new SpanFactory() {
                @Nullable
                @Override
                public Object getSpans(@NonNull MarkwonConfiguration configuration, @NonNull RenderProps props) {
                    final Object span = original.getSpans(configuration, props);
                    if (span instanceof AsyncDrawableSpan) {
                        ((AsyncDrawableSpan) span).getDrawable()
                                .setResult(/*YOU PLACEHOLDER DRAWABLE*/);
                    }
                    return span;
                }
            });
        }
    }
})
```

**Comment by noties** [comment]
Released in `3.0.0` version. Now it's possible to customize placeholder via image loading configuration
